qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -257,20 +323,90 @@ def add_data(
             if not self._service:
                 self._set_service_from_backend(self._backend)
 
-            self._jobs[job.job_id()] = job
+            if jid in self._jobs:
+                LOG.warning(
+                    "Skipping duplicate job, a job with this ID already exists [Job ID: %s]", jid
+                )
+            else:
+                self._jobs[jid] = job
+                if jid in self._job_futures:
+                    LOG.warning("Job future has already been submitted [Job ID: %s]", jid)
+                else:
+                    self._add_job_future(job)
+                    if timeout is not None:
+                        timeout_ids.append(jid)
 
-        if job_data:
-            job_kwargs = {
-                "jobs": job_data,
-                "timeout": timeout,
-            }
-            self._job_futures.append(
-                (job_kwargs, self._executor.submit(self._add_jobs_data, **job_kwargs))
-            )
+        # Add future for cancelling jobs that timeout
+        if timeout_ids:
+            self._job_executor.submit(self._timeout_running_jobs, timeout_ids, timeout)
 
         if self.auto_save:
             self.save_metadata()
 
+    def _timeout_running_jobs(self, job_ids, timeout):
+        """Function for cancelling jobs after timeout length.
+
+        This function should be called by an executor.
+
+        Args:
+            job_ids: the IDs of jobs to wait for.
+            timeout: The total time to wait for all jobs before cancelling.
+        """
+        futs = [self._job_futures[jid] for jid in job_ids]
+        waited = futures.wait(futs, timeout=timeout)
+
+        # Try to cancel timed-out jobs
+        if waited.not_done:
+            LOG.debug("Cancelling running jobs that exceeded add_jobs timeout.")
+            done_ids = {fut.result()[0] for fut in waited.done}
+            notdone_ids = [jid for jid in job_ids if jid not in done_ids]
```

Comment:
Not done futures won't have a result, the job id is only returned when the future finishes. This is why it can be extracted only from the done futures and the reason for the above work around to figure out the not-done job ids.